rust/sdk-processor/src/steps/parquet_default_processor/gcs_handler.rs

```diff
@@ -0,0 +1,185 @@
+use crate::parquet_processors::{ParquetTypeEnum, ParquetTypeStructs};
+use anyhow::Context;
+use aptos_indexer_processor_sdk::utils::errors::ProcessorError;
+use async_trait::async_trait;
+use google_cloud_storage::client::Client as GCSClient;
+use parquet::{
+    file::{properties::WriterProperties, writer::SerializedFileWriter},
+    record::RecordWriter,
+    schema::types::Type,
+};
+use processor::bq_analytics::{
+    gcs_handler::upload_parquet_to_gcs,
+    generic_parquet_processor::{GetTimeStamp, HasParquetSchema, HasVersion},
+};
+use std::{collections::HashMap, path::PathBuf, sync::Arc};
+use tracing::{debug, error};
+
+pub struct GCSUploader {
+    gcs_client: Arc<GCSClient>,
+    parquet_type_to_schemas: HashMap<ParquetTypeEnum, Arc<Type>>,
+    parquet_type_to_writer: HashMap<ParquetTypeEnum, SerializedFileWriter<Vec<u8>>>,
+    pub bucket_name: String,
+    pub bucket_root: String,
+    pub processor_name: String,
+}
+
+#[async_trait]
+pub trait Uploadable {
+    async fn handle_buffer(
+        &mut self,
+        buffer: ParquetTypeStructs,
+    ) -> anyhow::Result<(), ProcessorError>;
+}
+
+#[async_trait]
+impl Uploadable for GCSUploader {
+    async fn handle_buffer(
+        &mut self,
+        buffer: ParquetTypeStructs,
+    ) -> anyhow::Result<(), ProcessorError> {
+        // Directly call `upload_generic` for each buffer type
+        let table_name = buffer.get_table_name();
+
+        let result = match buffer {
+            ParquetTypeStructs::Transaction(transactions) => {
+                self.upload_generic(&transactions[..], ParquetTypeEnum::Transaction, table_name)
+                    .await
+            },
+            ParquetTypeStructs::MoveResource(resources) => {
+                self.upload_generic(&resources[..], ParquetTypeEnum::MoveResource, table_name)
+                    .await
+            },
+            ParquetTypeStructs::WriteSetChange(changes) => {
+                self.upload_generic(&changes[..], ParquetTypeEnum::WriteSetChange, table_name)
+                    .await
+            },
+            ParquetTypeStructs::TableItem(items) => {
+                self.upload_generic(&items[..], ParquetTypeEnum::TableItem, table_name)
+                    .await
+            },
+            ParquetTypeStructs::MoveModule(modules) => {
+                self.upload_generic(&modules[..], ParquetTypeEnum::MoveModule, table_name)
+                    .await
+            },
+        };
+
+        if let Err(e) = result {
+            error!("Failed to upload buffer: {}", e);
+            return Err(ProcessorError::ProcessError {
+                message: format!("Failed to upload buffer: {}", e),
+            });
+        }
+        Ok(())
+    }
+}
+
+pub fn create_new_writer(schema: Arc<Type>) -> anyhow::Result<SerializedFileWriter<Vec<u8>>> {
```

Comment:
does this need to be pub?

Reply:
yes it needs to be used when we initialize parquet buffer step